archival: Ensure that skip_offset_range only appears in reupload mode

oleiman opened the pull request:
During segment collection for upload, a leadership change can occur, leaving the uploader fiber to complete any work in progress before going to sleep. So the following is possible:

1. Segment collection for new segment upload on S0, the leader
2. Partition leadership changes to S1 before (1) completes
3. S1 uploads the next segment, updates manifest
3. (1) continues to the point of checking its upload candidate against the manifest to see whether a compacted upload would be a no-op. Normally the lookup fails in new_upload mode because the segment does not appear in the manifest yet. However in this case the manifest was already updated elsewhere, so the lookup succeeds, effectively tricking segment_collector into thinking it is in reupload mode. If the size of the collected segment matches what S1 already uploaded, then the candidate is discarded via skip_offset_range.
4. Back at the call site in ntp_archiver_service, skip_offset_range is expected to appear only in reupload mode, but this was a new upload. This triggers an assert.

This commit guards against that situation by plumbing the upload mode into make_segment_upload and adds a check to the reupload size sanity check to ensure that we are indeed in reupload mode before returning skip_offset_range.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
c/a/segment_collector: Add segment_collector_mode to constructor

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
c/a/segment_collector: Remove mode from collect_segments interface

And perform all operations based on this->_mode.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 1

oleiman commented:
Worth noting that the kind of clunky plumbing of the segment collector mode is improved in a different outstanding commit, where the mode is moved on the `segment_collector` instance. We can make that change here, but it seems fine as a followup.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--3d6ada28-a3ca-45eb-abc1-007cad30d401-->
<!--TEST RESULTS-->
<!--f1edb358-3247-4f2c-83cc-5f83a02540ce-->
<!--TEST RESULTS-->
<!--7ba7c250-ff82-453e-800e-714ecfb45764-->
<!--TEST RESULTS-->
<!--e47de98e-d142-462b-84e6-95acafff596d-->
<!--TEST RESULTS-->
<!--eed72eb8-72cc-45aa-b823-345b4cfdcfe5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67347">build#67347</a></summary>


| test_class                                   | test_method                                            | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------------------------|:-------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest                     | test_node_operations                                   | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67347#01976b2a-e24f-4720-909c-1ae3df341970 | FLAKY         | 20/21    | upstream reliability is '98.11320754716981'. current run reliability is '95.23809523809523'. drift is 2.87511 and the allowed drift is set to 50. The test should PASS |
| TxUpgradeTest                                | upgrade_does_not_change_tx_coordinator_assignment_test |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67347#01976b2a-e250-4a8d-a617-765e6ab459f0 | FLAKY         | 19/21    | upstream reliability is '93.84615384615384'. current run reliability is '90.47619047619048'. drift is 3.36996 and the allowed drift is set to 50. The test should PASS |
| src/v/crypto/tests/crypto_bench_rpbench_test | src/v/crypto/tests/crypto_bench_rpbench_test           |                                                                                                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/67347#01976af0-21e4-440d-85aa-d6ef276d9960 | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67356">build#67356</a></summary>


| test_class                               | test_method                         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-----------------------------------------|:------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| ControllerLogLimitPartitionBalancerTests | test_partition_balancer_with_limits |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67356#01976c78-d319-49a3-91fc-c9e4089ef003 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67896">build#67896</a></summary>


| test_class                 | test_method                              | test_arguments          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:---------------------------|:-----------------------------------------|:------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest      | test_fuzz_admin_ops                      |                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67896#0197a82a-8a1b-47f5-9f9d-76ca765a66b7 | FLAKY         | 19/21    | upstream reliability is '96.55172413793103'. current run reliability is '90.47619047619048'. drift is 6.07553 and the allowed drift is set to 50. The test should PASS  |
| DisablingPartitionsTest    | test_disable                             |                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67896#0197a829-a835-4828-9cf9-f489c4e6ba2b | FLAKY         | 17/21    | upstream reliability is '92.74611398963731'. current run reliability is '80.95238095238095'. drift is 11.79373 and the allowed drift is set to 50. The test should PASS |
| TxAtomicProduceConsumeTest | test_basic_tx_consumer_transform_produce | {"with_failures": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67896#0197a829-a835-466f-99c1-4824849bb32d | FLAKY         | 20/21    | upstream reliability is '99.66329966329967'. current run reliability is '95.23809523809523'. drift is 4.4252 and the allowed drift is set to 50. The test should PASS   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67936">build#67936</a></summary>


| test_class        | test_method          | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:------------------|:---------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| TopicRecoveryTest | test_missing_segment | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67936#0197a97c-7e78-4600-a569-4507900fccdd | FLAKY         | 31/40    | upstream reliability is '93.57429718875501'. current run reliability is '62.5'. drift is 31.0743 and the allowed drift is set to 50. The test should PASS               |
| TopicRecoveryTest | test_missing_segment | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67936#0197a97c-7e79-4985-9fa4-1311e9f3ce83 | FLAKY         | 37/40    | upstream reliability is '93.57429718875501'. current run reliability is '85.71428571428571'. drift is 7.86001 and the allowed drift is set to 50. The test should PASS  |
| TopicRecoveryTest | test_missing_segment | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67936#0197a97c-7e79-4cc2-b4b2-3f0ae8de4920 | FLAKY         | 32/40    | upstream reliability is '93.57429718875501'. current run reliability is '66.66666666666666'. drift is 26.90763 and the allowed drift is set to 50. The test should PASS |
| TopicRecoveryTest | test_missing_segment | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67936#0197a97c-7e7a-4218-a5da-938994046497 | FLAKY         | 33/40    | upstream reliability is '92.27722772277228'. current run reliability is '70.83333333333334'. drift is 21.44389 and the allowed drift is set to 50. The test should PASS |
| TopicRecoveryTest | test_missing_segment | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67936#0197a97c-7e7a-4602-9f47-af2d0b323d3b | FLAKY         | 35/40    | upstream reliability is '92.3076923076923'. current run reliability is '78.26086956521739'. drift is 14.04682 and the allowed drift is set to 50. The test should PASS  |
| TopicRecoveryTest | test_missing_segment | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67936#0197a97c-7e7a-46fa-bf37-197c3cb22aeb | FLAKY         | 36/40    | upstream reliability is '92.27722772277228'. current run reliability is '83.33333333333334'. drift is 8.94389 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68028">build#68028</a></summary>


| test_class                       | test_method                | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------------|:---------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDelayedTranslationTest   | test_basic                 | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "trino"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68028#0197ae47-7553-4ec5-8465-a38e38d7bba8 | FLAKY         | 20/21    | upstream reliability is '99.68152866242038'. current run reliability is '95.23809523809523'. drift is 4.44343 and the allowed drift is set to 50. The test should PASS |
| NodePostRestartProbeTest         | post_restart_probe_test    |                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68028#0197ae47-7b03-41b2-9a68-772ec94a1aa4 | FLAKY         | 20/21    | upstream reliability is '96.05911330049261'. current run reliability is '95.23809523809523'. drift is 0.82102 and the allowed drift is set to 50. The test should PASS |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss |                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68028#0197ae47-7b03-41b2-9a68-772ec94a1aa4 | FLAKY         | 20/21    | upstream reliability is '99.14772727272727'. current run reliability is '95.23809523809523'. drift is 3.90963 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

oleiman commented:
/ci-repeat 1
debug

oleiman committed:
c/a/segment_collector: Ensure that skip_offset_range only in reupload mode

    During segment collection for upload, a leadership change can occur,
    leaving the uploader fiber to complete any work in progress before
    going to sleep. So the following is possible:

    1. Segment collection for new segment upload on S0, the leader
    2. Partition leadership changes to S1 before (1) completes
    3. S1 uploads the next segment, updates manifest
    3. (1) continues to the point of checking its upload candidate
       against the manifest to see whether a compacted upload would
       be a no-op. Normally the lookup fails in new_upload mode because
       the segment does not appear in the manifest yet. However in this
       case the manifest was already updated elsewhere, so the lookup
       succeeds, effectively tricking segment_collector into thinking
       it is in reupload mode. If the size of the collected segment
       matches what S1 already uploaded, then the candidate is discarded
       via skip_offset_range.
    4. Back at the call site in ntp_archiver_service, skip_offset_range
       is expected to appear only in reupload mode, but this was a new
       upload. This triggers an assert.

    This commit guards against that situation by plumbing the upload
    mode into make_segment_upload and adds a check to the reupload size
    sanity check to ensure that we are indeed in reupload mode before
    returning skip_offset_range.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 1
release
skip-redpanda-build
skip-units
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.test_missing_segment
dt-repeat 30

oleiman commented:
/ci-repeat 1
release
arch=arm64
skip-redpanda-build
skip-units
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.test_missing_segment
dt-repeat=30

oleiman commented:
/ci-repeat 3
release
arch=arm64
skip-units
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.test_missing_segment
dt-repeat=20

oleiman commented:
/ci-repeat 1
skip-redpanda-build
skip-units

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68022-->
<!--74344235-0f3f-49c0-b836-6e5f66ede08f-->
<!--128723f5-1f83-4011-8235-dae0e5be9056-->
<!--45ff24d0-5266-4715-aae9-e46260e456e6-->
<!--366d6d3e-be4a-4935-88de-d221b17c599e-->
<!--a70d8917-e71d-4035-ace5-2c16b3eeb763-->
<!--c38dceb0-7a86-437d-8cc6-460791b9c685-->
<!--6e39852e-9a82-4e86-8304-8b05b79d2f7d-->
<!--title-->
## Retry command for [Build#68022](https://buildkite.com/redpanda/redpanda/builds/68022)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/schema_registry_test.py::SchemaRegistryACLTest.test_acl_idempotency
tests/rptest/tests/schema_registry_test.py::SchemaRegistryACLTest.test_acl_delete_filters
tests/rptest/tests/schema_registry_test.py::SchemaRegistryACLTest.test_acl_get_filters
tests/rptest/tests/schema_registry_test.py::SchemaRegistryACLTest.test_basic_acl_operations
tests/rptest/tests/schema_registry_test.py::SchemaRegistryACLTest.test_acl_validation
tests/rptest/tests/schema_registry_test.py::SchemaRegistryACLTest.test_field_variations
tests/rptest/tests/schema_registry_test.py::SchemaRegistryACLTest.test_case_handling
```
<!--/body-->

oleiman commented:
/ci-repeat 1

nvartolomei commented on src/v/cluster/archival/segment_reupload.cc:
```diff
@@ -1014,6 +1021,13 @@ ss::future<candidate_creation_result> segment_collector::make_upload_candidate(
+              archival_log.debug,
+              "New segment upload already appears in manifest: {}",
+              _segments.front());
+            co_return candidate_creation_error::concurrency_error;
```
This is fine and avoids the problem at hand but I think for non-reupload mode we shouldn't even drop into the to_replace check.

The check is still useful as in case someone updates the manifest concurrently we don't try to do an useless upload. However, if the new leader didn't yet update the manifest, from what I'm reading, we would still do the useless upload which we won't be able to replicate anyway.

Instead, there should be some sort of abort source present somewhere which would abort the operations/in-progress upload when we give away the leadership.

wdyt?

oleiman replied:
I don't disagree; sounds like a worthy optimization. I have a fair amount of archival code moving around on various branches, so I'll have to punt on that a little bit. I can put it on my TODO list though, certainly.
